luaui/Widgets/gui_spectator_hud.lua

```diff
@@ -0,0 +1,1949 @@
+	titleDimensions.verticalCenterOffset = math.floor(titleDimensions.height / 2)
+end
+
+local function calculateKnobDimensions()
+	knobDimensions.fontSize = math.floor(defaults.knobDimensions.fontSize * scaleMultiplier)
+	knobDimensions.padding = math.floor(defaults.knobDimensions.padding * scaleMultiplier)
+
+	knobDimensions.height = metricDimensions.height - 2 * knobDimensions.padding
+	knobDimensions.width = knobDimensions.height * constants.knobDimensions.heightToWidthFactor
+
+	knobDimensions.cornerSize = math.floor(defaults.knobDimensions.cornerSize * scaleMultiplier)
+	knobDimensions.outline = math.floor(defaults.knobDimensions.outline * scaleMultiplier)
+
+	knobDimensions.leftKnobLeft = titleDimensions.right + knobDimensions.padding
+	knobDimensions.leftKnobRight = knobDimensions.leftKnobLeft + knobDimensions.width
+
+	knobDimensions.rightKnobRight = widgetDimensions.right - widgetDimensions.borderPadding - knobDimensions.padding
+	knobDimensions.rightKnobLeft = knobDimensions.rightKnobRight - knobDimensions.width
+end
+
+local function calculateBarDimensions()
+	barDimensions.padding = math.floor(defaults.barDimensions.padding * scaleMultiplier)
+	barDimensions.paddingFromMetric = knobDimensions.padding + barDimensions.padding
+
+	local barHeight = knobDimensions.height - 2 * barDimensions.padding
+	local lineHeight = defaults.barDimensions.lineHeight * scaleMultiplier
+	barDimensions.lineMiddleOffset = math.floor((barHeight - lineHeight) / 2)
+
+	barDimensions.left = knobDimensions.leftKnobRight + 1
+	barDimensions.right = knobDimensions.rightKnobLeft - 1
+
+	barDimensions.width = barDimensions.right - barDimensions.left - knobDimensions.width
+end
+
+local function calculateDimensions()
+	calculateWidgetDimensions()
+	calculateMetricDimensions()
+	calculateTitleDimensions()
+	calculateKnobDimensions()
+	calculateBarDimensions()
+end
+
+local function updateMetricTextTooltips()
+	if WG['tooltip'] then
+		for metricIndex,metric in ipairs(metricsEnabled) do
+			local bottom = widgetDimensions.top - metricIndex * metricDimensions.height
+			local top = bottom + metricDimensions.height
+
+			local left = titleDimensions.left
+			local right = titleDimensions.right
+
+			WG['tooltip'].AddTooltip(
+				string.format("spectator_hud_vsmode_%d", metric.id),
+				{ left, bottom, right, top },
+				metric.tooltip,
+				nil,
+				metric.title
+			)
+		end
+	end
+end
+
+local function initMovingAverage(movingAverage)
+	if not settings.useMovingAverage then
+		movingAverage.average = 0
+		return
+	end
+
+	movingAverage.average = 0
+	movingAverage.index = 0
+	movingAverage.data = {}
+	for i=1,settings.movingAverageWindowSize do
+		movingAverage.data[i] = 0
+	end
+end
+
+local function updateMovingAverage(movingAverage, newValue)
+	if not settings.useMovingAverage then
+		movingAverage.average = newValue
+		return
+	end
+
+	if movingAverage.index == 0 then
+		for i=1,settings.movingAverageWindowSize do
```

Comment:
For the record, you dont actually need recalc this with a loop:
newavg = (oldavg * (settings.movingAverageWindowSize -1 ) + newvalue) / settings.movingAverageWindowSize

Reply:
Oh boy, this is implemented so long ago that I no longer remember the exact reason for it :/ This should definitely have a comment explaining what's going on because it's far from obvious. Allow me to explain the code.

In Lua, indexes start from 1. The moving average code here initializes the index to 0 which is a special value. When updating the moving average for the first time with a value, all of the window is set to the value (i.e. the first observed value).

The normal case where we just update a single value has no looping. We just replace the oldest value with the newest value. See line 816 with the code:

```
	movingAverage.average = movingAverage.average + (newValue - oldValue) / settings.movingAverageWindowSize
```

I think the reason for populating the window with the first encountered value is because of the scenario where the widget is enabled mid game. If the window was initialized with all zero's, the widget would show increasing stats for everything until the moving average window is filled with non-bogus values.